Index: src/InstanceScripts/Raid_BlackTemple.cpp
===================================================================
--- src/InstanceScripts/Raid_BlackTemple.cpp	(revision 853)
+++ src/InstanceScripts/Raid_BlackTemple.cpp	(working copy)
@@ -148,7 +148,7 @@
 		AuraPointer pAura = AuraPointer(new Aura(dbcSpell.LookupEntry(MUTANT_WAR_HOUND_CLOUD_OF_DISEASE),(uint32)20000, _unit, _unit));
 		if (pAura != NULL)
 		{
-			_unit->AddAura(pAura);
+			_unit->AddAura(pAura, NULLAURA);
 		}
 	}
 };
@@ -3216,7 +3216,7 @@
 			AuraPointer aura = AuraPointer(new Aura(dbcSpell.LookupEntry(SpellId),(uint32)15000, _unit, _unit));
 			if (aura)
 			{
-				_unit->AddAura(aura);
+				_unit->AddAura(aura, NULLAURA);
 			}
 
 			AuraChange = t + 15;
Index: src/InstanceScripts/Raid_CoT_BattleOfMountHyjal.cpp
===================================================================
--- src/InstanceScripts/Raid_CoT_BattleOfMountHyjal.cpp	(revision 853)
+++ src/InstanceScripts/Raid_CoT_BattleOfMountHyjal.cpp	(working copy)
@@ -541,7 +541,7 @@
 						AuraPointer aura = AuraPointer(new Aura(spells[1].info, (uint32)5000, _unit, _unit));
 						if (aura != NULL)
 						{
-							_unit->AddAura(aura);
+							_unit->AddAura(aura, NULLAURA);
 						}
 					}
 
@@ -1093,7 +1093,7 @@
 				AuraPointer aura = AuraPointer(new Aura(spells[3].info, (uint32)20000, _unit, RTarget));
 				if (aura != NULL)
 				{
-					RTarget->AddAura(aura);
+					RTarget->AddAura(aura, NULLAURA);
 				}
 
 				TargetTable.clear();
Index: src/InstanceScripts/Raid_Magtheridons_Lair.cpp
===================================================================
--- src/InstanceScripts/Raid_Magtheridons_Lair.cpp	(revision 853)
+++ src/InstanceScripts/Raid_Magtheridons_Lair.cpp	(working copy)
@@ -251,7 +251,7 @@
 							AuraPointer aura = AuraPointer(new Aura(dbcSpell.LookupEntry(SOUL_TRANSFER), (uint32)-1, BuffedChanneler, BuffedChanneler));
 							if (aura != NULLAURA)
 							{
-								BuffedChanneler->AddAura(aura);
+								BuffedChanneler->AddAura(aura, NULLAURA);
 							}
 						}
 					}
@@ -618,7 +618,7 @@
 			AuraPointer aura = AuraPointer(new Aura(dbcSpell.LookupEntry(MIND_EXHAUSTION),(uint32)78000, Magtheridon, Channeler));
 			if (aura != NULL)
 			{
-				Channeler->AddAura(aura);
+				Channeler->AddAura(aura, NULLAURA);
 			}
 
 			MagYell = true;
@@ -1349,7 +1349,7 @@
 		AuraPointer aura = AuraPointer(new Aura(dbcSpell.LookupEntry(BANISHMENT), (uint32)-1, _unit, _unit));
 		if (aura != NULL)
 		{
-			_unit->AddAura(aura);
+			_unit->AddAura(aura, NULLAURA);
 		}
 		_unit->CastSpell(_unit, dbcSpell.LookupEntry(BANISH), true);
 		_unit->GetAIInterface()->SetAllowedToEnterCombat(false);
Index: src/InstanceScripts/Raid_Naxxramas.cpp
===================================================================
--- src/InstanceScripts/Raid_Naxxramas.cpp	(revision 853)
+++ src/InstanceScripts/Raid_Naxxramas.cpp	(working copy)
@@ -409,7 +409,7 @@
 			// I don't like the way we apply it
 			AuraPointer WidowEmbrace( new Aura( dbcSpell.LookupEntry( NAXXRAMAS_WORSHIPPER_WIDOW_EMBRACE ), 30000, _unit, mGrandWidow->GetUnit() ) );
 			if ( WidowEmbrace != NULLAURA )
-				_unit->AddAura( WidowEmbrace );
+				_unit->AddAura( WidowEmbrace, NULLAURA );
 
 			// Not sure about new Frenzy Timer
 			mGrandWidow->ResetTimer( mGrandWidow->mFrenzyTimer, 60000 + RandomUInt( 20 ) * 1000 );
