Index: projects/InstanceScripts2008.vcproj
===================================================================
--- projects/InstanceScripts2008.vcproj	(revision 703)
+++ projects/InstanceScripts2008.vcproj	(working copy)
@@ -1,7 +1,7 @@
 <?xml version="1.0" encoding="windows-1250"?>
 <VisualStudioProject
 	ProjectType="Visual C++"
-	Version="9.00"
+	Version="9,00"
 	Name="SunInstanceScripts"
 	ProjectGUID="{EE39B48B-C598-4AAF-9EF0-199402933EE5}"
 	Keyword="Win32Proj"
@@ -102,9 +102,9 @@
 			/>
 		</Configuration>
 		<Configuration
-			Name="Release|Win32"
-			OutputDirectory="..\..\..\bin\release\script_bin"
-			IntermediateDirectory="2003_int_release_Suninstancescripts"
+			Name="Debug|x64"
+			OutputDirectory="..\..\..\bin\debug_x64\script_bin"
+			IntermediateDirectory="2008_int_debug_Suninstancescripts_x64"
 			ConfigurationType="2"
 			InheritedPropertySheets="$(VCInstallDir)VCProjectDefaults\UpgradeFromVC71.vsprops"
 			CharacterSet="2"
@@ -123,13 +123,18 @@
 			/>
 			<Tool
 				Name="VCMIDLTool"
+				TargetEnvironment="3"
 			/>
 			<Tool
 				Name="VCCLCompilerTool"
 				AdditionalOptions="/MP"
+				Optimization="0"
+				InlineFunctionExpansion="2"
 				AdditionalIncludeDirectories="..\..\hearthstone-shared;..\..\hearthstone-world;..\..\..\dep\include;..\..\..\dep\src"
-				PreprocessorDefinitions="WIN32;NDEBUG;_WINDOWS;_USRDLL;SCRIPTLIB"
-				RuntimeLibrary="0"
+				PreprocessorDefinitions="WIN32;_DEBUG;_WINDOWS;_USRDLL;SCRIPTLIB;X64;_CRT_SECURE_NO_WARNINGS;_CRT_NONSTDC_NO_WARNINGS"
+				MinimalRebuild="true"
+				BasicRuntimeChecks="3"
+				RuntimeLibrary="1"
 				UsePrecompiledHeader="0"
 				WarningLevel="3"
 				Detect64BitPortabilityProblems="false"
@@ -147,18 +152,16 @@
 			<Tool
 				Name="VCLinkerTool"
 				AdditionalDependencies="hearthstone-world.lib"
-				OutputFile="../../../bin/release/script_bin/SunInstanceScripts.dll"
-				LinkIncremental="1"
-				AdditionalLibraryDirectories="..\..\..\bin\release"
+				OutputFile="../../../bin/debug_x64/script_bin/SunInstanceScripts.dll"
+				LinkIncremental="2"
+				AdditionalLibraryDirectories="..\..\..\bin\debug_x64"
 				GenerateDebugInformation="true"
-				ProgramDatabaseFile="../../../bin/release/script_bin/SunInstanceScripts.pdb"
+				ProgramDatabaseFile="../../../bin/debug_x64/script_bin/SunInstanceScripts.pdb"
 				SubSystem="2"
-				OptimizeReferences="2"
-				EnableCOMDATFolding="2"
 				RandomizedBaseAddress="1"
 				DataExecutionPrevention="0"
 				ImportLibrary="$(OutDir)/SunInstanceScripts.lib"
-				TargetMachine="1"
+				TargetMachine="17"
 			/>
 			<Tool
 				Name="VCALinkTool"
@@ -183,9 +186,9 @@
 			/>
 		</Configuration>
 		<Configuration
-			Name="Debug|x64"
-			OutputDirectory="..\..\..\bin\debug_x64\script_bin"
-			IntermediateDirectory="2008_int_debug_Suninstancescripts_x64"
+			Name="Release|Win32"
+			OutputDirectory="..\..\..\bin\release\script_bin"
+			IntermediateDirectory="2003_int_release_Suninstancescripts"
 			ConfigurationType="2"
 			InheritedPropertySheets="$(VCInstallDir)VCProjectDefaults\UpgradeFromVC71.vsprops"
 			CharacterSet="2"
@@ -204,18 +207,13 @@
 			/>
 			<Tool
 				Name="VCMIDLTool"
-				TargetEnvironment="3"
 			/>
 			<Tool
 				Name="VCCLCompilerTool"
 				AdditionalOptions="/MP"
-				Optimization="0"
-				InlineFunctionExpansion="2"
 				AdditionalIncludeDirectories="..\..\hearthstone-shared;..\..\hearthstone-world;..\..\..\dep\include;..\..\..\dep\src"
-				PreprocessorDefinitions="WIN32;_DEBUG;_WINDOWS;_USRDLL;SCRIPTLIB;X64;_CRT_SECURE_NO_WARNINGS;_CRT_NONSTDC_NO_WARNINGS"
-				MinimalRebuild="true"
-				BasicRuntimeChecks="3"
-				RuntimeLibrary="1"
+				PreprocessorDefinitions="WIN32;NDEBUG;_WINDOWS;_USRDLL;SCRIPTLIB"
+				RuntimeLibrary="0"
 				UsePrecompiledHeader="0"
 				WarningLevel="3"
 				Detect64BitPortabilityProblems="false"
@@ -233,16 +231,18 @@
 			<Tool
 				Name="VCLinkerTool"
 				AdditionalDependencies="hearthstone-world.lib"
-				OutputFile="../../../bin/debug_x64/script_bin/SunInstanceScripts.dll"
-				LinkIncremental="2"
-				AdditionalLibraryDirectories="..\..\..\bin\debug_x64"
+				OutputFile="../../../bin/release/script_bin/SunInstanceScripts.dll"
+				LinkIncremental="1"
+				AdditionalLibraryDirectories="..\..\..\bin\release"
 				GenerateDebugInformation="true"
-				ProgramDatabaseFile="../../../bin/debug_x64/script_bin/SunInstanceScripts.pdb"
+				ProgramDatabaseFile="../../../bin/release/script_bin/SunInstanceScripts.pdb"
 				SubSystem="2"
+				OptimizeReferences="2"
+				EnableCOMDATFolding="2"
 				RandomizedBaseAddress="1"
 				DataExecutionPrevention="0"
 				ImportLibrary="$(OutDir)/SunInstanceScripts.lib"
-				TargetMachine="17"
+				TargetMachine="1"
 			/>
 			<Tool
 				Name="VCALinkTool"
@@ -373,6 +373,22 @@
 				RelativePath="..\src\InstanceScripts\Setup.h"
 				>
 			</File>
+			<Filter
+				Name="Lang"
+				>
+				<File
+					RelativePath="..\src\InstanceScripts\Lang\Config.h"
+					>
+				</File>
+				<File
+					RelativePath="..\src\InstanceScripts\Lang\en_US.h"
+					>
+				</File>
+				<File
+					RelativePath="..\src\InstanceScripts\Lang\ru_RU.h"
+					>
+				</File>
+			</Filter>
 		</Filter>
 		<Filter
 			Name="Scripts"
Index: src/InstanceScripts/Base.h
===================================================================
--- src/InstanceScripts/Base.h	(revision 703)
+++ src/InstanceScripts/Base.h	(working copy)
@@ -22,7 +22,7 @@
 
 #include "StdAfx.h"
 #include "Setup.h"
-
+#include "Lang/Config.h"
 #include <string>
 #include <vector>
 #include <map>
Index: src/InstanceScripts/Lang/Config.h
===================================================================
--- src/InstanceScripts/Lang/Config.h	(revision 0)
+++ src/InstanceScripts/Lang/Config.h	(revision 0)
@@ -0,0 +1,8 @@
+#define En_US
+//#define ru_RU
+#ifdef En_US
+#include "En_US.h"
+#endif
+#ifdef ru_RU
+#include "ru_RU.h"
+#endif
\ No newline at end of file
Index: src/InstanceScripts/Lang/en_US.h
===================================================================
--- src/InstanceScripts/Lang/en_US.h	(revision 0)
+++ src/InstanceScripts/Lang/en_US.h	(revision 0)
@@ -0,0 +1 @@
+#define LUck "lkkn"
\ No newline at end of file
Index: src/InstanceScripts/Lang/ru_RU.h
===================================================================
